File: android-sdk/modelhitch-core/src/main/kotlin/com/genoventureslabs/modelhitch/Provider.kt
package com.genoventureslabs.modelhitch

import kotlinx.coroutines.flow.Flow

interface Provider {
    val id: String
    val name: String
    val defaultModel: String
    val capabilities: Capabilities

    suspend fun chat(
        params: ChatParams,
        credentials: ProviderCredentials = ProviderCredentials(),
    ): ChatResult

    fun stream(
        params: ChatParams,
        credentials: ProviderCredentials = ProviderCredentials(),
    ): Flow<StreamChunk>

    suspend fun listModels(
        credentials: ProviderCredentials = ProviderCredentials(),
    ): List<ModelInfo> = emptyList()
}

File: flutter-sdk/modelhitch_dart/lib/src/provider.dart
import 'types.dart';

abstract interface class Provider {
  String get id;
  String get name;
  String get defaultModel;
  Capabilities get capabilities;

  Future<ChatResult> chat(ChatParams params, ProviderCredentials credentials);
  Stream<StreamChunk> stream(
      ChatParams params, ProviderCredentials credentials);
  Future<List<ModelInfo>> listModels(ProviderCredentials credentials);
}
